### DISCUSSION TITLE
Implementation of the "Python array API standard" in Mojo?

### DISCUSSION THREAD

**Discussion by paugier** [modular/modular#1281] [Ideas]
It would be great to build a Mojo implementation of the [Python array API standard](https://data-apis.org/array-api/latest/), that would be usable both from Python (after [creation of a Python package with AOT compilation](https://github.com/modularml/mojo/discussions/1161)) and from Mojo.

This API standard has been designed to be friendly with tools relying on a static language. For example, the types never depends on the run time values (in contrast to Numpy).

I realize that it is a huge work but it could worth it from the point of view of Python interoperability and partial migration towards Mojo. Having such library would significantly help users to migrate their Python code using Numpy (and other libraries using this standard, at least [TensorFlow](https://www.tensorflow.org/), [PyTorch](https://pytorch.org/), [MXNet](https://numpy.mxnet.io/), [JAX](https://github.com/google/jax), [Dask](https://dask.org/) and [CuPy](https://cupy.chainer.org/)) towards Mojo.

```
# in Python
# to use the Numpy implementation https://numpy.org/devdocs/reference/array_api.html
# from numpy import array_api as xp
# or to use the Mojo implementation
# import mojo_array_api as xp
# in Mojo
import py_array_api as xp

arr = xp.arange(1000)
...
```

Of course, in Mojo the arrays would just be native Mojo objects (without interaction with the Python interpreter) so that Mojo can have the opportunity to obtain decent performance with such code.

Note that there is [a test suite](https://github.com/data-apis/array-api-tests) for array libraries adopting the Python Array API standard.

Note: of course a goal of "copy-paste in Mojo your Python array API standard code and it works" would be nice. ~~However, [I'm still puzzled with the difference of semantic of the assign operator]() (reference creation in Python and copy in Mojo) and I don't see how one can create a Mojo type which is not copied during an assignment (which might not be a good idea anyway).~~ The main `array` struct can be a reference to a container struct to mimic Python behavior (example [here](https://github.com/modularml/mojo/discussions/1276#discussioncomment-7569682))

Note: of course a good thing for this potential project is that Mojo would be a very good tool to implement this standard, in particular with `from tensor import Tensor, TensorSpec, TensorShape`.

**Comment by guidorice** [comment]
I referenced this discussion over here in a feature request: https://github.com/modularml/mojo/issues/1515

> memoryview builtin and support for python buffer protocol
